lib/spack/spack/environment/environment.py

```diff
@@ -1760,22 +1761,14 @@ def install_specs(self, specs=None, **install_args):
 
     def all_specs(self):
         """Return all specs, even those a user spec would shadow."""
-        all_specs = set()
-        for h in self.concretized_order:
-            try:
-                spec = self.specs_by_hash[h]
-            except KeyError:
-                tty.warn(
-                    "Environment %s appears to be corrupt: missing spec " '"%s"' % (self.name, h)
-                )
-                continue
-            all_specs.update(spec.traverse())
-
-        return sorted(all_specs)
+        roots = [self.specs_by_hash[h] for h in self.concretized_order]
```

Comment:
notice that we don't have this defensive validation of `KeyError` in tons of other places, so I dropped it here too.